core/watcher/eventsource/value.go

```diff
@@ -20,80 +20,111 @@ type ValueWatcher struct {
-) *ValueWatcher {
-	w := &ValueWatcher{
-		BaseWatcher: base,
-		out:         make(chan struct{}),
-		namespace:   namespace,
-		filter: func(e changestream.ChangeEvent) bool {
-			return e.Changed() == changeValue
-		},
-		changeMask: changeMask,
-		mapper:     mapper,
+// Namespace is the namespace to watch for changes.
+func (f filter) Namespace() string {
+	return f.namespace
+}
+
+// ChangeMask is the type of change to watch for.
+func (f filter) ChangeMask() changestream.ChangeType {
+	return f.changeMask
+}
+
+// ChangePredicate is a function that returns true if the change event is
+// for the desired value.
+func (f filter) ChangePredicate() func(string) bool {
+	return f.changePredicate
+}
+
+// ValueFilter returns a filter option that watches for changes in the namespace
+// that match the change mask and cause the supplied predicate to return true.
+func ValueFilter(namespace string, changeMask changestream.ChangeType, changePredicate func(string) bool) FilterOption {
+	return filter{
+		namespace:       namespace,
+		changeMask:      changeMask,
+		changePredicate: changePredicate,
 	}
+}
 
-	w.tomb.Go(w.loop)
-	return w
+// NamespaceFilter returns a filter option that watches for changes in the
+// namespace that match the change mask.
+func NamespaceFilter(namespace string, changeMask changestream.ChangeType) FilterOption {
+	return filter{
+		namespace:       namespace,
+		changeMask:      changeMask,
+		changePredicate: func(string) bool { return true },
+	}
 }
 
-// NewNamespaceNotifyWatcher returns a new watcher that receives changes from the input
-// base watcher's db/queue when changes in the namespace occur.
-func NewNamespaceNotifyWatcher(base *BaseWatcher, namespace string, changeMask changestream.ChangeType) *ValueWatcher {
-	return NewNamespaceNotifyMapperWatcher(base, namespace, changeMask, defaultMapper)
+// NewFilterWatcher returns a new watcher that receives changes from the
+// input base watcher's db/queue when change-log events occur for a specific
+// changeValue from the input namespace.
+func NewFilterWatcher(
+	base *BaseWatcher, filterOptions ...FilterOption,
+) *ValueWatcher {
+	return NewFilterMapperWatcher(base, defaultMapper, filterOptions...)
 }
 
-// NewNamespaceNotifyMapperWatcher returns a new watcher that receives changes from
-// the input base watcher's db/queue when changes in the namespace occur.
-func NewNamespaceNotifyMapperWatcher(
-	base *BaseWatcher, namespace string, changeMask changestream.ChangeType, mapper Mapper,
+// NewFilterMapperWatcher returns a new watcher that receives changes from
+// the input base watcher's db/queue when mapper accepts the change-log events
+// for a specific changeValue from the input namespace.
+func NewFilterMapperWatcher(
+	base *BaseWatcher, mapper Mapper, filterOptions ...FilterOption,
 ) *ValueWatcher {
+	opts := make([]changestream.SubscriptionOption, len(filterOptions))
+	for i, opt := range filterOptions {
+		predicate := opt.ChangePredicate()
```

Comment:
There's no guard that ChangePredicate can never be nil. `ValueFilter` call be called with a nil param and there's no check done.

Not as serious but along the same lines, namespace is not checked to ensure it's not ""

Reply:
If it's nil and you didn't add tests, I kinda think that's on you.

Reply:
My point is **this** package is missing tests and input validation. Each layer should not trust input passed to it. There's no easy way to test it in this package, but do this in the changestream eventmultiplexer package and watch a nice NPE

```
func (s *eventMultiplexerSuite) TestSubscribeWithNilFilter(c *gc.C) {
	s.testMultipleDispatch(c, changestream.FilteredNamespace("topic", changestreamtesting.Update, nil))
}
```

`FilteredNamespace` should guard against accepting a nil func.
As should the `ValueFilter` and `NamespaceFilter` constructors.